api/server/server.go

```diff
@@ -0,0 +1,156 @@
+package server
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+
+	"github.com/andres-teleport/overseer/api"
+	"github.com/andres-teleport/overseer/api/authentication"
+	"github.com/andres-teleport/overseer/lib/multipipe"
+	"github.com/andres-teleport/overseer/lib/supervisor"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	ErrEmptyCommand = status.Error(codes.InvalidArgument, "empty job command provided")
+)
+
+type Server struct {
+	jobOwners  map[string]string
+	mu         *sync.RWMutex
+	supervisor *supervisor.Supervisor
+	srv        *grpc.Server
+	l          net.Listener
+	api.UnimplementedJobworkerServiceServer
+}
+
+func NewServer(listenAddr, keyFile, certFile, caFile string) (*Server, error) {
+	creds, err := authentication.NewServerTransportCredentials(keyFile, certFile, caFile)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Server{
+		jobOwners:  make(map[string]string),
+		mu:         &sync.RWMutex{},
+		supervisor: supervisor.NewSupervisor(),
+	}
+
+	authInterceptor := NewAuthorizationInterceptor(s)
+
+	s.srv = grpc.NewServer(
+		grpc.Creds(creds),
+		grpc.UnaryInterceptor(authInterceptor.unaryServerInterceptor),
+		grpc.StreamInterceptor(authInterceptor.streamServerInterceptor),
+	)
+	api.RegisterJobworkerServiceServer(s.srv, s)
+
+	l, err := net.Listen("tcp", listenAddr)
```

Comment:
Might be useful to add a some sort of cleanup to the server which will close the listener.